
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Commands &mdash; VAMP 0.3.3.1 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.3.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="VAMP 0.3.3.1 documentation" href="index.html" />
    <link rel="next" title="Modules" href="modules.html" />
    <link rel="prev" title="Usage" href="usage.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="modules.html" title="Modules"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="usage.html" title="Usage"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">VAMP 0.3.3.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="commands">
<h1>Commands<a class="headerlink" href="#commands" title="Permalink to this headline">Â¶</a></h1>
<dl class="docutils" id="clean-multi-ssake-sh">
<dt><strong>clean_multi_ssake.sh</strong></dt>
<dd><p class="first">Run to remove intermediate files generated by multi_ssake.py from the current
directory.</p>
<p class="last"><em>WARNING</em>: Removes all files matching *.fasta and *paired.fa from the current
directory!</p>
</dd>
</dl>
<dl class="docutils" id="compare-genomes-py">
<dt><strong>compare_genomes.py</strong></dt>
<dd>Finds differences between genomes based on the input multi-aligned fasta file</dd>
</dl>
<div class="highlight-text"><div class="highlight"><pre>python: can&#39;t open file &#39;compare_genomes_test.py&#39;: [Errno 2] No such file or directory
</pre></div>
</div>
<dl class="docutils" id="fastq-to-fasta-py">
<dt><strong>fastq_to_fasta.py</strong></dt>
<dd>Converts fastq files to fasta files</dd>
</dl>
<div class="highlight-text"><div class="highlight"><pre>Usage: fastq_to_fasta.py [options] fastq_file fasta_file

 Convert fastq file to fasta

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  -w WRAP, --wrap=WRAP  Maximum length of lines, 0 means do not wrap (default:
                        0)
  -v, --verbose         verbose output
</pre></div>
</div>
<dl class="docutils" id="find-contig-deletions-py">
<dt><strong>find_contig_deletions.py</strong></dt>
<dd>Find contigs in an assembly that have sections deleted and provide an option to
insert the missing pieces</dd>
</dl>
<div class="highlight-text"><div class="highlight"><pre>Usage: fastq_to_fasta.py [options] fastq_file fasta_file

 Convert fastq file to fasta

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  -w WRAP, --wrap=WRAP  Maximum length of lines, 0 means do not wrap (default:
                        0)
  -v, --verbose         verbose output
</pre></div>
</div>
<dl class="docutils">
<dt><strong>gff2gtf_simple.py</strong></dt>
<dd>Attempts to convert GFF files to GTF files.</dd>
</dl>
<div class="highlight-text"><div class="highlight"><pre>Usage: gff2gtf_simple.py [options] gff_file

 Compares genomes using multiple alignment Input: GFF Output: GTF File  Very
simple and naieve GFF to GTF converter. Writen to handle GFF output from
BioPerls genbank2gff3.pl script for simple DNA viruses

Options:
  --version      show program&#39;s version number and exit
  -h, --help     show this help message and exit
  -v, --verbose  verbose output
</pre></div>
</div>
<dl class="docutils" id="maf-net-py">
<dt><strong>maf_net.py</strong></dt>
<dd>Stitch together alignment block from a MAF file, forming an alignment net.</dd>
</dl>
<div class="highlight-text"><div class="highlight"><pre>Usage: maf_net.py [options] maf_file

 Determine the best MAF block (determined by score) that cover a specified
genome

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  -r REFERENCE, --reference=REFERENCE
                        Reference species
  -s SPECIES, --species=SPECIES
                        List of species to include
  -c CHROMOSOME, --chromosome=CHROMOSOME
                        Sequence ID of the chromosome for which to generate
                        the alignment net (e.g. NC_001806)
  -o OUTPUT_DIR, --output_dir=OUTPUT_DIR
                        Directory to store output file, default is maf file
                        directory
  --consensus_sequence  Output &quot;consensus sequence&quot; for each species in files
                        named [species].[chromosome].consensus.fasta
  --reference_fasta=REFERENCE_FASTA
                        Check MAF file against this fasta (for
                        troubleshooting, debugging)
  -v, --verbose         verbose output
</pre></div>
</div>
<dl class="docutils" id="makepairedoutput2equalfiles-vamp-pl">
<dt><strong>makePairedOutput2EQUALfiles_vamp.pl</strong></dt>
<dd><p class="first">See <a class="reference internal" href="#makepairedoutput2unequalfiles-vamp-pl"><em>makePairedOutput2UNEQUALfiles_vamp.pl</em></a>:</p>
<div class="last highlight-python"><pre>Usage: makePairedOutput2EQUALfiles_vamp.pl &lt;fasta file 1&gt; &lt;fasta file 2&gt; &lt;library insert size&gt;
       --- ** Both files must have the same number of records &amp; arranged in the same order</pre>
</div>
</dd>
</dl>
<dl class="docutils" id="makepairedoutput2unequalfiles-vamp-pl">
<span id="index-0"></span><dt><strong>makePairedOutput2UNEQUALfiles_vamp.pl</strong></dt>
<dd><p class="first">Modified versions of scripts provided by <a class="reference external" href="http://www.bcgsc.ca/platform/bioinfo/software/ssake">SSAKE</a>.  They are used to prepare two
separate paired end fastq files for use by <a class="reference external" href="http://www.bcgsc.ca/platform/bioinfo/software/ssake">SSAKE</a>.  The modifications made were
to accommodate new
<a class="reference external" href="http://en.wikipedia.org/wiki/FASTQ_format#Illumina_sequence_identifiers">Illumina style sequence identifiers</a>
introduced with CASAVA 1.8.:</p>
<div class="last highlight-python"><pre>Usage: makePairedOutput2UNEQUALfiles_vamp.pl &lt;fasta file 1&gt; &lt;fasta file 2&gt; &lt;library insert size&gt;
       --- files could have different # of records &amp; arranged in different order but template ids must match</pre>
</div>
</dd>
</dl>
<dl class="docutils" id="multi-ssake-py">
<dt><strong>multi_ssake.py</strong></dt>
<dd>Run TQSFastq preprocessing and <a class="reference external" href="http://www.bcgsc.ca/platform/bioinfo/software/ssake">SSAKE</a> using various combinations of parameters and combine the results.</dd>
</dl>
<div class="highlight-text"><div class="highlight"><pre>Usage: multi_ssake.py [options] forward_reads reverse_reads

 Run various iterations of SSAKE, varying input files and parameters Collect
results into single list of contigs

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  -i INSERT_SIZE, --insert_size=INSERT_SIZE
                        Mean insert size for paired reads (default: 500)
  --config=CONFIG       multi_ssake configuration file that specifies options
                        [default: /Users/lparsons/Documents/workspace_4.2/vamp
                        /makefiles/multi_ssake.config.template]
  --vamp_config=VAMP_CONFIG
                        vamp configuration file that specifies options
                        [default: /Users/lparsons/Documents/workspace_4.2/vamp
                        /makefiles/config.mk.template]
  --qsub                Use qsub to submit commands to cluster (default:
                        False)
  --untrimmed           Include SSAKE assembly of untrimmed reads (not
                        recommended)
  -v, --verbose         verbose output
  -d, --debug           debug (do not execute)
</pre></div>
</div>
<dl class="docutils">
<dt><strong>translate_cds.py</strong></dt>
<dd>Takes GFF and Fasta input and translates spliced CDS regions from DNA to Amino Acid sequence, reporting errors.</dd>
</dl>
<div class="highlight-text"><div class="highlight"><pre>Usage: translate_cds.py [options] &lt;genes gff3&gt; &lt;ref fasta&gt;

 Extracts coding sequences (cds) regions from fasta reference and gff file and
translates them into amino acid sequence, output in FASTA format to STDOUT.
Errors during translation are output to STDERR. Genes with translation errors
are not printed.

Options:
  --version             show program&#39;s version number and exit
  -h, --help            show this help message and exit
  --notrans             Do not translate to amino acid sequence, output DNA
  -i IDATTR, --idattr=IDATTR
                        GFF attribute to use as gene ID. Features with the
                        same ID will be considered parts of the same gene. The
                        default &quot;gene_id&quot; is suitable for GTF files.
  -t FEATURETYPE, --featuretype=FEATURETYPE
                        GFF feature type(s) (3rd column) to be used. Specify
                        the option multiple times for multiple feature types.
                        The default is &quot;CDS&quot; for GFF files and &quot;CDS&quot; and
                        &quot;stop_codon&quot; for GTF files.
  --table=TABLE         NCBI Translation table to use when translating DNA
                        (see http://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprint
                        gc.cgi). Default: 1.
  -v, --verbose         verbose output
</pre></div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="usage.html"
                        title="previous chapter">Usage</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="modules.html"
                        title="next chapter">Modules</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/command_index.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="modules.html" title="Modules"
             >next</a> |</li>
        <li class="right" >
          <a href="usage.html" title="Usage"
             >previous</a> |</li>
        <li><a href="index.html">VAMP 0.3.3.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Lance Parsons.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>