
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Usage &mdash; VAMP 0.3.3.1 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.3.3.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="VAMP 0.3.3.1 documentation" href="index.html" />
    <link rel="next" title="Commands" href="command_index.html" />
    <link rel="prev" title="Installation" href="install.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="command_index.html" title="Commands"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="install.html" title="Installation"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">VAMP 0.3.3.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h1>
<div class="section" id="preprocessing">
<h2>PreProcessing<a class="headerlink" href="#preprocessing" title="Permalink to this headline">¶</a></h2>
<p>This step removes artifacts and contaminate sequences from you data in an effort
to speed up and improve the quality of the resulting assemblies.</p>
<ol class="arabic">
<li><dl class="first docutils">
<dt>Run <tt class="docutils literal"><span class="pre">setup_preprocessing.py</span></tt></dt>
<dd><ul class="first last">
<li><dl class="first docutils">
<dt>You will be asked to provide</dt>
<dd><ul class="first last simple">
<li>Forward read file - The filename of a fastq file (gzipped is
ok) containing the forward reads from sequencing</li>
<li>Reverse read file - If you performed paired-end sequencing,
this is the filename of the fastq file containing the reverse read.
Note, that the reads must be in matching order to the forward read
file and the read ids should be the same (characters after a space in
the read id are ignored for matching purposes).</li>
<li>Output directory - The location where the processed files
will be stored</li>
<li>Output prefix - The basename of the output files (typically
the project or strain name)</li>
<li>Host reference - Either a FASTA file or the basename of a
Bowtie index. Sequences will be aligned to this reference and any
that match will be removed from the analysis. This is typically used
to remove host contamination.</li>
</ul>
</dd>
</dl>
</li>
<li><p class="first">The script will generate a makefile in the specified output
directory called <tt class="docutils literal"><span class="pre">PREFIX_preprocess.mk</span></tt>.</p>
</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Execute resulting makefile</dt>
<dd><ul class="first last simple">
<li>You can view the resulting makefile to see the various parameters
that will be used and edit any configuration options.</li>
<li>Running <tt class="docutils literal"><span class="pre">make</span> <span class="pre">-f</span> <span class="pre">PREFIX_preprocess.mk</span></tt> will execute the various
preprocessing steps. You make take advantage of multiple processors by
using the <tt class="docutils literal"><span class="pre">-j</span> <span class="pre">N</span></tt> options to specify the number of threads for <tt class="docutils literal"><span class="pre">make</span></tt>
to use.</li>
<li>To cleanup intermediate files: <tt class="docutils literal"><span class="pre">make</span> <span class="pre">-f</span> <span class="pre">analysis.mk</span> <span class="pre">clean</span></tt></li>
</ul>
</dd>
</dl>
</li>
</ol>
<p>The PreProcessing consists of the following steps. After each step, FastQC is
run to generate a summary of the result.</p>
<blockquote>
<div>1. Run <a class="reference external" href="https://github.com/marcelm/cutadapt">Cutadapt</a> to trim adapter
sequences from the ends of reads. (Intermediate, output is discarded). 2.
Run <a class="reference external" href="http://hannonlab.cshl.edu/fastx_toolkit/">Fastx Toolkit</a>&#8216;s artifact
filter to remove low complexity sequences. 3. If paired reads are used,
match up paired reads and set unpaired reads aside. 4. Map reads using
<a class="reference external" href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a> to the specified
host reference, keeping only those reads (or pairs) that did not map.</div></blockquote>
</div>
<div class="section" id="short-read-de-novo-assembly">
<h2>Short Read de novo Assembly<a class="headerlink" href="#short-read-de-novo-assembly" title="Permalink to this headline">¶</a></h2>
<p>This step uses the
<a class="reference external" href="http://www.bcgsc.ca/platform/bioinfo/software/ssake">SSAKE</a> short read
assembler to build contigs from the preprocessed reads. It does this by running
SSAKE with various parameter settings and combining the results.</p>
<p>1. <tt class="docutils literal"><span class="pre">multi_ssake.py</span></tt> 2. Run <em class="xref std std-ref">clean_multi_ssake.sh</em> to cleanup intermediate
fasta files (interactive, do NOT use qsub)</p>
<p><em>TODO - Additional Detail Needed</em></p>
</div>
<div class="section" id="contig-assembly">
<h2>Contig Assembly<a class="headerlink" href="#contig-assembly" title="Permalink to this headline">¶</a></h2>
<p>The resulting contigs from the previous SSAKE assemblies are often redundant and
can be stitched together using a long-read assembler. We have had success using
<a class="reference external" href="http://staden.sourceforge.net/manual/gap4_unix_toc.html">Gap4</a> from the
<a class="reference external" href="http://bowtie-bio.sourceforge.net/index.shtml">Staden</a> package.</p>
<ol class="arabic simple">
<li>Assemble <tt class="docutils literal"><span class="pre">ssake_combined.contigs</span></tt> (using Gap4 or other long read assembler)</li>
</ol>
<p><em>TODO - Additional Detail Needed</em></p>
</div>
<div class="section" id="reference-guided-scaffold-assembly">
<h2>Reference Guided Scaffold Assembly<a class="headerlink" href="#reference-guided-scaffold-assembly" title="Permalink to this headline">¶</a></h2>
<p>Once you have generated the best contigs you can with
<a class="reference external" href="http://www.bcgsc.ca/platform/bioinfo/software/ssake">SSAKE</a> and
<a class="reference external" href="http://staden.sourceforge.net/manual/gap4_unix_toc.html">Gap4</a>, the next
step is to put them together into a scaffold. To do this, VAMP utilizes the
reference genome of a similar strain to determine how to order the contigs.</p>
<div class="section" id="align-contigs-to-reference">
<h3>Align Contigs to Reference<a class="headerlink" href="#align-contigs-to-reference" title="Permalink to this headline">¶</a></h3>
<p>The first step is to align contigs to a reference genome and output the result
in a <a class="reference external" href="http://genome.ucsc.edu/FAQ/FAQformat.html#format5">MAF</a> formatted file.
There are many options for alignment tools, however, we have had success with
<a class="reference external" href="http://mugsy.sourceforge.net/">Mugsy</a>, a very fast multiple whole genome
alignment tool.</p>
</div>
<div class="section" id="stitch-together-scaffold">
<h3>Stitch Together Scaffold<a class="headerlink" href="#stitch-together-scaffold" title="Permalink to this headline">¶</a></h3>
<p>Once you have aligned the contigs to the reference, the next step is to stitch
together the various alignment blocks into a scaffold.  The <em class="xref std std-ref">maf_net.py</em>
utility does this by reassembling the reference sequence from the MAF blocks
and using the highest scoring block for each location in the genome to assemble
a scaffold genome.</p>
<dl class="docutils">
<dt>Output:</dt>
<dd><ul class="first last simple">
<li>Aligned Fasta File - <tt class="docutils literal"><span class="pre">BASENAME.net.afa</span></tt> This file contains an
aligned fasta file created by stitching together MAF blocks based on the
reference sequence.  Where two blocks overlap, the higher scoring block
is used.</li>
</ul>
</dd>
<dt>Optional Output (one per species):</dt>
<dd><ul class="first last">
<li><p class="first">Consensus Sequence - <tt class="docutils literal"><span class="pre">SPECIES.consensus.fasta</span></tt> A FASTA file
containing the consensus sequence for this species. N&#8217;s in the sequence
represent sections where no contigs mapped to a section of the reference
(i.e. potential gaps in the scaffold).</p>
</li>
<li><p class="first">Consensus Contig Composition GFF -
<tt class="docutils literal"><span class="pre">SPECIES.consensus_contig_composition.gff</span></tt> GFF formatted file describes
intervals in the SPECIES genome. The attributes contain information about
the contigs used to determine the sequence in this interval.  The
attributes are:</p>
<blockquote>
<div><ul class="simple">
<li>src_seq</li>
<li>src_seq_start</li>
<li>src_seq_end</li>
<li>src_strand</li>
<li>src_size</li>
<li>maf_block</li>
<li>block_start</li>
<li>block_end</li>
<li>ref_src</li>
<li>ref_start</li>
<li>ref_end</li>
<li>ref_strand</li>
</ul>
</div></blockquote>
</li>
<li><p class="first">Consensus Contig Composition Summary -
<tt class="docutils literal"><span class="pre">SPECIES.consensus_contig_composition_summary.txt</span></tt> Tab delimited file
with the following columns that describes intervals in the SPECIES genome
and the contigs that were used for the sequence.</p>
<blockquote>
<div><ul class="simple">
<li>seq - sequence id of the interval in the SPECIES
genome</li>
<li>start - start position of the interval</li>
<li>end - end position of the interval</li>
<li>contig - contig id that was used to &#8220;build&#8221; this
interval.  If <em>None</em>, that means no contig was found for the
analagous region in the reference.</li>
<li>contig_start - the start position of the contig that
aligned to this start interval</li>
<li>contig_end - the end position of the contig that
aligned to the end position of this interval</li>
<li>contig_strand - the direction that the contig
aligned to the reference (if &#8216;-&#8216;, the reverse complement of the
contig aligned to the reference in this interval)</li>
<li>contig_size - the full size of the contig (including
those bases that did not aligned to this interval)</li>
</ul>
</div></blockquote>
</li>
</ul>
</dd>
</dl>
</div>
</div>
<div class="section" id="genome-annotation-and-comparison">
<h2>Genome Annotation and Comparison<a class="headerlink" href="#genome-annotation-and-comparison" title="Permalink to this headline">¶</a></h2>
<p>Once a draft of a genome has been completed, it can be useful to migrate
annotations from an annotated reference to the new genome.  In addition, this
step generates a summary of the changes at the nucleic acid as well as amino
acid level.</p>
<p>Run <em class="xref std std-ref">compare_genomes.py</em> to migrate annotations and generate a list of
differences between two species. The script requires an aligned fasta file
(typically use the one generated from the previous scaffold stitching step) and
a GFF file of features (genes, exons, etc.) to migrate.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Usage</a><ul>
<li><a class="reference internal" href="#preprocessing">PreProcessing</a></li>
<li><a class="reference internal" href="#short-read-de-novo-assembly">Short Read de novo Assembly</a></li>
<li><a class="reference internal" href="#contig-assembly">Contig Assembly</a></li>
<li><a class="reference internal" href="#reference-guided-scaffold-assembly">Reference Guided Scaffold Assembly</a><ul>
<li><a class="reference internal" href="#align-contigs-to-reference">Align Contigs to Reference</a></li>
<li><a class="reference internal" href="#stitch-together-scaffold">Stitch Together Scaffold</a></li>
</ul>
</li>
<li><a class="reference internal" href="#genome-annotation-and-comparison">Genome Annotation and Comparison</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="install.html"
                        title="previous chapter">Installation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="command_index.html"
                        title="next chapter">Commands</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/usage.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="command_index.html" title="Commands"
             >next</a> |</li>
        <li class="right" >
          <a href="install.html" title="Installation"
             >previous</a> |</li>
        <li><a href="index.html">VAMP 0.3.3.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Lance Parsons.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
  </body>
</html>